fix: guest link when the bbb activity lives under a group

keevan opened the pull request:
- adjust input to div which fixes default display of extra parameters as &amp;
- pass through the group id as part of the guest link
- add handling to ensure it resolves as expected on the other end also

keevan committed:
fix: guest link when the bbb activity lives under a group

- adjust input to div which fixes default display of extra parameters as &amp;
- pass through the group id as part of the guest link
- add handling to ensure it resolves as expected on the other end also

keevan commented on yui/src/rooms/js/rooms.js:
```diff
@@ -75,9 +75,7 @@ M.mod_bigbluebuttonbn.rooms = {
-                copyText.select();
-                copyText.setSelectionRange(0, 99999); /*For mobile devices*/
-                document.execCommand("copy");
+                navigator.clipboard.writeText(copyText.textContent);
```
execCommand is deprecated in favour of using the clipboard api directly:

https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand
https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/writeText

Opted to use writeText as it has better browser support (and the write method wouldn't work in firefox)

Peterburnett commented on classes/form/guestlink_access_form.php:
```diff
@@ -42,13 +42,11 @@ class guestlink_access_form extends \moodleform {
-        $mform->setType('guestlinkurl', PARAM_TEXT);
-        $mform->setDefault('guestlinkurl', $this->_customdata['guestlinkurl']);
+        $guestlinkelement = \html_writer::div($this->_customdata['guestlinkurl'], 'form-control readonly', ['id' => 'id_guestlinkurl']);
+        $inviteparticipantselement = $mform->createElement('static', 'guestlinkurl', get_string('view_guestlink_label', 'bigbluebuttonbn'), $guestlinkelement);
```
This should likely be a HTML element, but it might dork with the column positioning

keevan replied:
It was actually fine after all, this might have been due to the oddly no longer redundant `mform->addGroup` a little further down.

Peterburnett commented on guestlink.php:
```diff
@@ -110,6 +116,12 @@
     \mod_bigbluebuttonbn\locallib\bigbluebutton::view_bbbsession_set($context, $bbbsession);
+
+    // Groups handling
+    if ($group !== -1) {
```
I wonder if we should do some additional validation here. Can someone force a submission with a groupID they dont have access to, and get in?

keevan replied:
Since there doesn't appear to be a separate guest link generated per group (e.g. with a different url, access code, etc) it's using a single guest link with a group designation. Since these group ids are typically sequential in nature, it's easy to jump around if you know what you are doing.

A better solution might be to have a separate guest link available per group. This however might limit the possibility of a guest joining a different session if that was infact intended at any given point, and would mean the data would need to probably live in a separate table (one to many) since the groups and linking are generated based on base meeting id, and moodle's group id. Alternatively, since they are not both stored in the bigbluebuttonbn table, it might be necessary to add a `group` field to this table to allow for it to be unique and can be directly referenced.

tl;dr: It's a bit weak yes, depends if you think by design it should work one way or the other, or be configurable. Also, this would require a DB table change of some sort.

Peterburnett replied:
Id like to do a permissions check then, to see if the user can actually access this group. There should be a function in grouplib.php hopefully. Lets just spend a little bit of time to see if we can harden it a little

Peterburnett commented:
Are there any possible migration paths we could do in an upgrade to fix up these links that are currently broken? Im not sure we have the data, its just worth exploring if it is a possibility.

keevan commented:
> Are there any possible migration paths we could do in an upgrade to fix up these links that are currently broken? Im not sure we have the data, its just worth exploring if it is a possibility.

This is mentioned in more detail in the previous comment https://github.com/catalyst/moodle-mod_bigbluebuttonbn/pull/65#discussion_r931690656, but currently, it's generated at run time (combination of meetingid and moodle's group id) so the links aren't broken per se, and this update should allow the guest links shown to be fixed (to a group, if required) when loaded.

keevan committed:
fix: handling for BBB grouped activities for guests

- add a groupsalt field to use as the salt for the guest links in group mode
- ensure the name of the group is shown when the guest link is visited. This is visual confirmation that it works as expected.
- the added salt means different groups will have a different hash
- adjusted guestlink url input style and mform type
